build.d

```diff
@@ -0,0 +1,196 @@
+
+// Flags for DMD
+immutable OutputFlag = "-of" ~ DubBinPath;
+immutable IncludeFlag = "-I" ~ RootPath.buildPath("source");
+
+
+/// Entry point
+int main(string[] args)
+{
+    // This does not have a 'proper' CLI interface, as it's only used in
+    // special cases (e.g. package maintainers can use it for bootstrapping),
+    // not for general / everyday usage by newcomer.
+    // So the following is just an heuristic / best effort approach.
+    if (args.length > 2 ||
+        (args.length == 2 && (args[1].canFind("help", "?") || args[1] == "-h")))
+    {
+        writeln("USAGE: ", args[0], " [version]");
+        writeln("  In order to build DUB, a version module must first be generated.");
+        writeln("  If one is already existing, it won't be overriden. " ~
+                "Otherwise this script will use the first argument, if any, " ~
+                "or the GITVER environment variable.");
+        writeln("  If both are empty, `git describe` will be called");
+        writeln("  Build flags can be provided via the `DFLAGS` environment variable.");
+        writeln("  LDC or GDC can be used by setting the `DMD` value to " ~
+                "`ldmd2` and `gdmd` (or their path), respectively.");
+        return 1;
+    }
+
+    immutable dubVersion = args.length > 1 ? args[1] : environment.get("GITVER", "");
+    if (!writeVersionFile(dubVersion))
+        return 1;
+
+    immutable dmd = getCompiler();
+    if (!dmd.length) return 1;
+    immutable dflags = environment.get("DFLAGS", "-g -O -w").split();
+
+    // Compiler says no to immutable (because it can't handle the appending)
+    const command = [
+        dmd,
+        OutputFlag, IncludeFlag,
+        "-version=DubUseCurl", "-version=DubApplication",
+        ] ~ dflags ~ [ "@build-files.txt" ];
+
+    writeln("Building dub using ", dmd, ", this may take a while...");
+    auto proc = execute(command);
+    if (proc.status != 0)
+    {
+        writeln("Command `", command, "` failed, output was:");
+        writeln(proc.output);
+        return 1;
+    }
+
+    // Check dub
+    auto check = execute([DubBinPath, "--version"]);
+    if (check.status != 0)
+    {
+        writeln("Running newly built `dub` failed: ", check.output);
+        return 1;
+    }
+
+    writeln("DUB has been built as: ", DubBinPath);
+    version (Posix)
+        writeln("You may want to run `sudo ln -s ", DubBinPath, " /usr/local/bin` now");
+    else version (Windows)
+        writeln("You may want to add the following entry to your PATH " ~
+                "environment variable: ", DubBinPath);
+    return 0;
+}
+
+/**
+ * Generate the version file describing DUB's version / commit
+ *
+ * Params:
+ *   dubVersion = User provided version file. Can be `null` / empty,
+ *                in which case the existing file (if any) takes precedence,
+ *                or the version is infered with `git describe`.
+ *                A non-empty parameter will always override the existing file.
```

Comment:
Nit: official D style is without leading stars and it's also about consistency within the file (see too).

Reply:
Right, I'll fix.

On this point: `dub` doesn't follow the official D style *at all*. Be it in the placement of brackets, the usage of tabs, etc... 
What would be the acceptable way to deal with this:
1) Keep it as is (provided the style is not consistent, and under-documented);
2) Gradually improve it when a piece of code is worked on;
3) Improve it all at once;

I'm not fan of 1) because it confuses contributors, and 3 makes `git blame` a bit harder (but not impossible, especially since Github support the "blame before this revision" feature).

Reply:
Fixed style